src/wp-includes/media.php

```diff
@@ -5564,6 +5676,135 @@ function wp_get_loading_attr_default( $context ) {
 	return 'lazy';
 }
 
+/**
+ * Get loading optimization attributes.
+ *
+ * @since 6.3.0
+ *
+ * @param string $tag_name The tag name.
+ * @param array[] $attr Array of the attributes for the tag.
+ * @param string $context Context for the element for which the loading optimization attribute is requested.
+ *
+ * @return array[] loading optimization attributes.
+ */
+function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+	global $wp_query;
+
+	$loading_attrs = array();
+
+	// All three parameters are required.
+	if ( empty( $tag_name ) || empty( $attr ) || empty( $context ) ) {
+		return $loading_attrs;
+	}
+
+	// Image tag must have a minimum of width and height.
+	if ( 'img' === $tag_name && ( ! isset( $attr['width'] ) || ! isset( $attr['height'] ) ) ) {
+		return $loading_attrs;
+	}
+
+	// Skip lazy-loading for the overall block template, as it is handled more granularly.
+	// The skip is also applicable for fetchpriority.
+	if ( 'template' === $context ) {
+		return $loading_attrs;
+	}
+
+	// If the count has already reached or exceeded the threshold, all of the images after can be lazy-loaded.
+	$content_media_count = wp_increase_content_media_count( 0 );
+	if ( wp_omit_loading_attr_threshold() < $content_media_count ) {
+		$loading_attrs['loading'] = 'lazy';
+		return $loading_attrs;
+	}
+
+	$img_size = $attr['width'] * $attr['height'];
+
```

Comment:
Something that is generally missing from this function but we need to add is support for when a developer manually passed attributes. We have to consider those as well and not overwrite them. Such a condition could be added here for example.
```suggestion
	// If a 'loading' attribute is already provided, do not overwrite it.
	if ( isset( $attr['loading'] ) ) {
		if ( 'lazy' === $attr['loading'] ) {
			$loading_attrs['loading'] = 'lazy';
		}
		return wp_maybe_add_fetchpriority_high_attr( $loading_attrs, $attr );
	}

	// If a 'fetchpriority' attribute with value 'high' is already provided, do not add `loading="lazy"`.
	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
		return wp_maybe_add_fetchpriority_high_attr( $loading_attrs, $attr );
	}
```

Reply:
@felixarntz I have moved the check (with some change) above just after the `width` & `height` attribute check.

Reply:
@kt-12 The way you addressed this has a few problems compared to the solution I suggested above:
* For the first condition: If a `loading` attribute is already provided, we should still run it through the `wp_maybe_add_fetchpriority_high_attr()` function to potentially add `fetchpriority="high"`. The way you solved it the condition is only entered when `loading="lazy"` is passed, but that is a problem, e.g. when `loading="eager"` is passed we would still overwrite it, which shouldn't happen.
* For the second condition: We need to call `wp_maybe_add_fetchpriority_high_attr()` even if we know the `fetchpriority="high"` attribute is added, so that the function can count the image and set the flag to add the attribute to `false`. If someone manually marks the image as a priority, we need to respect that and then we shouldn't potentially add the attribute to another image that comes later.

Reply:
Thank you for pointing this out. I missed this point of view. I have fixed the code.